back-end/utils/top-layer/UserUtil.js

```diff
@@ -362,6 +362,173 @@ const getLengthUserTransactionsHistoryForRedisM5RU = (email, filters) => {
   });
 };
 
+/**
+ * Perform buy action for the function proceedTransaction
+ * @param {import(".prisma/client").User} user
+ * @param {number} totalPendingPrice
+ * @param {string} companyCode
+ * @returns
+ */
+ const buyShareEvent = async (
+  user,
+  totalPendingPrice,
+  companyCode,
+  quantity,
+  recentPrice
+) => {
+  try {
+    if (user.cash < totalPendingPrice) {
+      return new Error(
+        "The user does not have enough money to buy the pending stock"
+      );
+    }
+
+    user.cash -= totalPendingPrice;
+    const buyShare = user.shares.find(
+      (share) => share.companyCode === companyCode
+    );
+    const buyShareIndex = user.shares.indexOf(buyShare);
+
+    if (!buyShare) {
+      await prisma.share.create({
+        companyCode: companyCode,
+        quantity: quantity,
+        buyPriceAvg: recentPrice,
+        user: user,
+        userID: user.id
+      });
+    } else {
+      // Update in Share
+      buyShare.buyPriceAvg =
+        (buyShare.buyPriceAvg * buyShare.quantity +
+          recentPrice * (buyShare.quantity + quantity)) /
+        (quantity + buyShare.quantity);
+      buyShare.quantity += quantity;
+      await prisma.share.update({
+        where: {
+          id: buyShare.id
+        },
+        data: {
+          buyPriceAvg: buyShare.buyPriceAvg,
+          quantity: buyShare.quantity
+        }
+      });
+
+      // Update in User
+      user.shares[buyShareIndex] = buyShare;
+      await prisma.user.update({
+        where: {
+          id: user.id
+        },
+        data: {
+          shares: user.shares
+        }
+      });
+    }
+  } catch (err) {
+    console.error(err);
+  }
+};
+
+const sellShareEvent = async (
+  user,
+  totalPendingPrice,
+  companyCode,
+  quantity,
+  recentPrice
+) => {
+  try {
+    user.cash += totalPendingPrice;
+    const sellShare = user.shares.find(
+      (share) => share.companyCode === companyCode
+    );
+    const sellShareIndex = user.shares.indexOf(sellShare);
+
+    if (!sellShare) {
```

Comment:
Add a condition to check whether user's cash is sufficient:
`user.cash + totalCashChange < 0`